### DISCUSSION TITLE
Stride and WPF

### DISCUSSION THREAD

**Discussion by jhimes144** [stride3d/stride#2211] [Q&A] [answered]
Hello,

I'm exploring using WPF as UI for my game. I was able to get an app running with WPF and the game rendered inside. However, the game is being rendered on top of the UI. I've been trying to scour the source code to see how the Stride Editor does it, because its able to draw context menus on top of the game view. Any ideas? Thanks.

EDIT:

After messing around, I was able to get wpf rendered on top of the game by making a borderless wpf window that allowed transparency, and using win32 functions to sync its position with the game. However, this requires the game to run in windowed mode (borderless and fullscreen, but IsFullscreen is off), and in window mode the game's performance suffers a lot from excessive screen tearing, I've yet to find how to fix this.

**Comment by Doprez** [comment]
For your frame rate issue can you try adding this within your gamecontext?
```
WindowMinimumUpdateRate.MinimumElapsedTime = TimeSpan.FromMilliseconds(1000f / 144);
MinimizedMinimumUpdateRate.MinimumElapsedTime = TimeSpan.FromMilliseconds(1000f / 144);
```

It may be a weird issue where the window thinks its minimized or out of focus so it uses the default fps limit.

If you need a spot to put this code you can override the Game class like below:
```
public class CustomGame : Game
{
	protected override void BeginRun()
	{
		base.BeginRun();
                // force 144FPS when in focus
		WindowMinimumUpdateRate.MinimumElapsedTime = TimeSpan.FromMilliseconds(1000f / 144f);
                // force 144FPS when out of focus
		MinimizedMinimumUpdateRate.MinimumElapsedTime = TimeSpan.FromMilliseconds(1000f / 144f);
	}
}
```

and then in you startup where you run `game.Run()` use the `CustomGame` class instead of the `Game` class.

**Reply by jhimes144** [reply]
Thanks for the help, unfortunately I'm still getting screen tearing. (This is with the fps template btw)

**Reply by Doprez** [reply]
interesting last thing I know about is https://forums.stride3d.net/t/how-to-disable-frame-rate-throttle-and-vsync/2094 where it might be a vsync issue? if you try changing the `GraphicsDevice.Presenter.PresentInterval` to something else maybe it will help.

If this doesnt help would you be able to share what you have done so far with the window renderer? Maybe I or someone more knowledgable than me can play around with the code to see if something comes up.

**Reply by jhimes144** [reply]
I came a across that forum earlier and tried those options, thanks for the help. I just opened this issue https://github.com/stride3d/stride/issues/2216

**Reply by Doprez** [reply, marked as answer]
https://github.com/stride3d/stride/issues/2216#issuecomment-2040345384

can you mark this as the answer if it seems fixed now?

**Comment by luca-domenichini** [comment]
Hello everyone,

May I ask @jhimes144 how you did display WPF controls on top of Stride game?
I tried to do some experiments in this repo https://github.com/luca-domenichini/StrideEmbedding but unfortunately, in that WPF project I have a button that is rendered behind Stride game.

Could you give me a hint? (or a PR on that repo would be much appreciated! 🚀 )

**Reply by jhimes144** [reply]
I used win32 windowing apis. I ended up switching to Avalonia, and created this library to do all of the work. If you want to do the same with WPF, it should be the same win32 apis but the wpf side would be different. Avalonia is extremely similar to WPF though.

https://github.com/jhimes144/AvaStride

**Reply by luca-domenichini** [reply]
@jhimes144 Thanks for quick reply. Are you telling me that AvaStride is suitable to run Stride inside Avalonia? Reading from main description, it seemed to me that its purpose was to run Avalonia inside Stride and not the opposite... maybe I missed something.
Could you clarify?

**Reply by luca-domenichini** [reply]
Btw, I am evaluating both WPF and Avalonia right now... trying to pick the right-fit for next proj...

**Reply by luca-domenichini** [reply]
@jhimes144 maybe you could drop a comment here also... https://gist.github.com/westonsoftware/a3fa982397fe1817ece4a27d3cbc5a89

**Reply by jhimes144** [reply]
AvaStride runs Avalonia "inside" stride i.e it embeds Avalonia into stride. Running stride inside Avalonia or Wpf is more difficult. Its easy enough to get Stride to run embedded into wpf or avalonia (same tech as AvaStride but in reverse), but its more work to render controls on top of stride in this configuration. Last time I checked (which was probably a year ago) Stride's editor does this by creating 3 windows. The first window is the editor app (wpf window), the second is Stride (stride window), and the third is another wpf window, used to draw controls "on top of stride", but its all parlor tricks. These windows are then made frameless and children of each other through win32 api calls. I recall the Stride Editor code had a lot of hacks to get everything to play nice with click and keyboard events.

You cannot render controls on top of Stride when embedding Stride into WPF or Avalonia without using windowing techniques, unless you want to fork Stride and WPF or Avalonia code and hook directly into rendering process for each. I would not recommend this path, and I don't think anyone has done it this way. You would be working with low level OpenGL, DirectX and/or Skia code.

If you want to embed stride into avalonia, and be able to draw controls - then I would recommend re-purposing my code here https://github.com/jhimes144/AvaStride/blob/main/AvaStride/NativeHelper.cs. This includes bindings ready to go to call the needed win32 apis. You can see where I'm calling them here https://github.com/jhimes144/AvaStride/blob/main/AvaStride/AvaloniaInStride.cs. The actual windowing code is not that verbose. Most of the code in there is just code for talking back and forth between Stride and Avalonia threads. Using that code, I would recreate something similar to what Stride's Editor does.

**Reply by jhimes144** [reply]
The code in that gist looks like its doing a CPU bound texture copy of Stride into a bitmap to be rendered into Avalonia. This is also a workable option. However, this suffers from poor performance. I attempted this first for my project and didn't get the performance I needed.

**Reply by Doprez** [reply]
You have to be careful as to what you are looking for here. The solution Jhimes had was to render a separate application on top of the SDL/winforms implementation that stride renders in. It looks like you are trying to render into one of the 2 windowing frameworks (Avalonia/WPF).

You can instead render the SDL window using its handle as I have [here](https://github.com/Doprez/avalonia-hosted-stride). This does not solve the [airspace issue](https://github.com/AvaloniaUI/Avalonia/discussions/13953) of rendering Avalonia buttons on top of the game view which was the reason I started looking into the source with a custom Stride branch https://github.com/stride3d/stride/pull/2404 seen in my [repo example](https://github.com/Doprez/custom-stride-window-poc/tree/avalonia-custom-platform) using that branch. This seems to work but was a hacky solution so I am looking into something more concrete [here](https://github.com/stride3d/stride/pull/2574) but I still have lots to do before its ready.

You can also take a look at the code in the Stride editor that renders the GameWindow into the WPF scene view.
https://github.com/stride3d/stride/blob/581cef98c2dc87b4587b3d90590477bd36ab5bb6/sources/presentation/Stride.Core.Presentation.Wpf/Controls/GameEngineHost.cs
https://github.com/stride3d/stride/blob/581cef98c2dc87b4587b3d90590477bd36ab5bb6/sources/presentation/Stride.Core.Presentation.Wpf/Controls/GameEngineHost.cs

**Reply by jhimes144** [reply]
@Doprez 

Just to clarify, the solution doesn't run as a "separate application" as a separate process. Its all in one process. The code that "talks" between Avalonia and Stride is just code that dispatches between the two threads. No interprocess work is being done because everything is in the same process.

**Reply by Doprez** [reply]
Ah my mistake sorry, I completely misunderstood how your code was working in that case. I made that assumption based on the fullscreen issue but I guess that would also make sense just with it relying on the window and nothing to do with separate processes. Thank you for clarifying!
